### PR TITLE
[SPARK-41572][SQL] Assign name to _LEGACY_ERROR_TEMP_2149

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to assign name to _LEGACY_ERROR_TEMP_2149, "MALFORMED_CSV_RECORD".

### Why are the changes needed?

We should assign proper name to _LEGACY_ERROR_TEMP_*

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Fix UT & run `./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite*`

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 1]
```diff
@@ -370,8 +370,11 @@ abstract class CSVSuite
           .load(testFile(carsFile)).collect()
       }
 
-      assert(exception.getMessage.contains("Malformed CSV record"))
-      assert(ExceptionUtils.getRootCause(exception).isInstanceOf[RuntimeException])
+      checkError(
+        exception = ExceptionUtils.getRootCause(exception).asInstanceOf[SparkRuntimeException],
+        errorClass = "MALFORMED_CSV_RECORD",
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:375]: hmm... there is `_LEGACY_ERROR_TEMP_2177` on the error chain, should we handle it in this one together?

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -851,6 +851,11 @@
       "Cannot name the managed table as <identifier>, as its associated location <location> already exists. Please pick a different table name, or remove the existing location first."
     ]
   },
+  "MALFORMED_CSV_RECORD" : {
+    "message" : [
+      "Malformed CSV record"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:856]: Can we provide more info to users? At least which record is malformed, and where is the problem - schema mismatch, the converter failed, and so on.

Comment (itholic) [reply core/src/main/resources/error/error-classes.json]: Thanks!
I think we can provide the schema information, since it raised when schema mismatch.
Just add some more informations to error message.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 1 above)
Comment (itholic) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:381]: Just added test case for `_LEGACY_ERROR_TEMP_2177`.
Maybe we can assign the proper name for it in the separate ticket

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -851,6 +851,11 @@
       "Cannot name the managed table as <identifier>, as its associated location <location> already exists. Please pick a different table name, or remove the existing location first."
     ]
   },
+  "MALFORMED_CSV_RECORD" : {
+    "message" : [
+      "Malformed CSV record. The number of tokens <tokenLength> doesn't match the schema <schemaLength>."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:856]: Can we output the malformed CSV record? If an user process billions of CSV lines, how he/she should figure out which one is bad?

Comment (itholic) [reply core/src/main/resources/error/error-classes.json:856]: Seems like this exception is triggered when checking the schema before parsing the actual input row.

Maybe should we fix the error class name and message more properly?

For example:
```json
  "MALFORMED_CSV_SCHEMA" : {
    "message" : [
      "Malformed CSV schema. The number of tokens <tokenLength> doesn't match the schema <schemaLength>."
    ]
  }
```

Or maybe we can show the malformed schema itself:
```json
  "MALFORMED_CSV_SCHEMA" : {
    "message" : [
      "Malformed CSV schema. <malformedSchema>."
    ]
  }
```

This will display look like:
```
[MALFORMED_CSV_RECORD] Malformed CSV schema: StructType(StructField(year,StringType,true),StructField(make,StringType,true),StructField(model,StringType,true),StructField(comment,StringType,true),StructField(blank,StringType,true))
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 4]
```diff
@@ -370,8 +370,17 @@ abstract class CSVSuite
           .load(testFile(carsFile)).collect()
       }
 
-      assert(exception.getMessage.contains("Malformed CSV record"))
-      assert(ExceptionUtils.getRootCause(exception).isInstanceOf[RuntimeException])
+      checkError(
+        exception = exception.getCause.asInstanceOf[SparkException],
+        errorClass = "_LEGACY_ERROR_TEMP_2177",
+        parameters = Map("failFastMode" -> "FAILFAST")
+      )
```
Comment (itholic) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:377]: This fails on `CSVv2Suite` since it raises `_LEGACY_ERROR_TEMP_2064` instead of `_LEGACY_ERROR_TEMP_2177`.

Is there any way we separate the tests between CSVv1 and CSVv2 ??

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:856]: This one process actual CSV records:
```
  private def convert(tokens: Array[String]): Option[InternalRow] = {
    if (tokens == null) {
      throw BadRecordException(
        () => getCurrentInput,
        () => None,
        QueryExecutionErrors.malformedCSVRecordError())
    }
```
and `getCurrentInput` can get you a bad record. Can't it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:377]: You could move the check to those specific test suites.

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (itholic) [reply core/src/main/resources/error/error-classes.json:856]: Oh... I didn't know `getCurrentInput`. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 5]
```diff
@@ -319,15 +319,17 @@ class UnivocityParser(
       throw BadRecordException(
         () => getCurrentInput,
         () => None,
-        QueryExecutionErrors.malformedCSVRecordError())
+        QueryExecutionErrors.malformedCSVRecordError(""))
     }
 
+    val currentInput = getCurrentInput
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:325]: It is not used in regular cases, correct? Don't think we should introduce additional overhead. Please, use `getCurrentInput` directly in errors.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 6]
```diff
@@ -3138,13 +3141,54 @@ class CSVv1Suite extends CSVSuite {
     super
       .sparkConf
       .set(SQLConf.USE_V1_SOURCE_LIST, "csv")
+
+  private val carsFile = "test-data/cars.csv"
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3145]: The same is defined in the parent class (just make it as `protected`). Please, remove it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 7]
```diff
@@ -3138,13 +3141,54 @@ class CSVv1Suite extends CSVSuite {
...
+        spark.read
+          .format("csv")
+          .option("multiLine", multiLine)
+          .options(Map("header" -> "true", "mode" -> "failfast"))
+          .load(testFile(carsFile)).collect()
+      }
+
+      checkError(
+        exception = exception.getCause.asInstanceOf[SparkException],
+        errorClass = "_LEGACY_ERROR_TEMP_2177",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3159]: Could you explain why did you add the test for the error class in the PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 8]
```diff
@@ -3138,13 +3141,54 @@ class CSVv1Suite extends CSVSuite {
...
+        spark.read
+          .format("csv")
+          .option("multiLine", multiLine)
+          .options(Map("header" -> "true", "mode" -> "failfast"))
+          .load(testFile(carsFile)).collect()
+      }
+
+      checkError(
+        exception = exception.getCause.asInstanceOf[SparkException],
+        errorClass = "_LEGACY_ERROR_TEMP_2064",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3186]: The same question like above. How this is related to assigning a name to `_LEGACY_ERROR_TEMP_2149`?

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 7 above)
Comment (itholic) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3159]: I covered it as well in this PR because it's in the same error chain. it is found by: https://github.com/apache/spark/pull/39258#discussion_r1058263385

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 8 above)
Comment (itholic) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:3186]: Yeah, it's because of the same reason with the above. Do we want to do it in separate PR and revert it from here?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
